spec/consensus/light-client.md

```diff
@@ -60,259 +61,368 @@ For the purpose of this lite client specification, we assume that the Tendermint
+  if h2, exists := Store.Get(height); exists {
+    if isWithinTrustedPeriod(h2) { return nil }
+    return ErrHeaderNotWithinTrustedPeriod(h2)
+  }
 
-_Verification Condition:_ We may need a Tendermint invariant stating that if _h2.Header.height = h1.Header.height + 1_ then _signers(h2.Commit) \subseteq h1.Header.NextV_.
+  h2 := Commit(height)
+  if !verify(h2) { return ErrInvalidHeader(h2) }
+  if !isWithinTrustedPeriod(h2) { return ErrHeaderNotWithinTrustedPeriod(h2) }
 
-_Remark_: The variable _trustlevel_ can be used if the user believes that relying on one correct validator is not sufficient. However, in case of (frequent) changes in the validator set, the higher the _trustlevel_ is chosen, the more unlikely it becomes that CheckSupport returns true for non-adjacent headers.
+  // get the highest trusted headers lower than h2
+  h1 = Store.HighestTrustedSmallerThan(height)
+  if h1 == nil { return ErrNoTrustedHeader }
 
-**Bisection.** The following function uses CheckSupport in a recursion to find intermediate headers that allow to establish a sequence of trust.
+  err = CanTrust(h1, h2, trustThreshold)  // or CanTrustBisection((h1, h2, trustThreshold)
+  if err != nil { return err }
 
-```go
-func Bisection(h1,h2,trustlevel) bool{
-  if CheckSupport(h1,h2,trustlevel) {
-    return true
+  if isWithinTrustedPeriod(h2) {
+    Store.add(h2)
+    // we store only trusted headers, as we assume that only trusted headers
+    // are influencing end user business decisions.
+    return nil
   }
-  if h2.Header.height == h1.Header.height + 1 {
-    // we have adjacent headers that are not matching (failed
-    // the CheckSupport)
-    // we could submit evidence here
-    return false
+  return ErrHeaderNotTrusted(h2)
+}
+
+
+// return nil in case we can trust header h2 based on header h1; otherwise return error
+// where error captures the nature of the error.
+func CanTrust(h1,h2,trustThreshold) error {
+  assume h1.Header.Height < h2.header.Height
+
+  th := h1 // th is trusted header
+  untrustedHeaders := [h2]
+
+  while true {
+    for h in untrustedHeaders {
+        // we assume here that iteration is done in the order of header heights
+        err = CheckSupport(th,h,trustThreshold)
+        if err == nil {
+            th = h
+            Store.Add(h)
+            untrustedHeaders.RemoveHeadersSmallerOrEqual(h.Header.Height)
+            if th == h2 { return nil }
+        }
+        if (err != ErrTooMuchChange) { return err }
+    }
+
+    endHeight = min(untrustedHeaders)
+    foundPivot = false
+    while(!foundPivot) {
+        pivot := (th.Header.height + endHeight) / 2
+        hp := Commit(pivot)
+        if !verify(hp) { return ErrInvalidHeader(hp) }
+        // try to move trusted header forward to hp
+        err = CheckSupport(th,hp,trustThreshold)
+        if (err != nil and err != ErrTooMuchChange) return err
+        if err == nil {
+            th = hp
+            Store.Add(hp)
+            foundPivot = true
```

Comment:
we need a break here so we don't add `th` to the `untrustedHeaders` below
```suggestion
            foundPivot = true
            break
```

Alternatively:
```
   ....
    endHeight = min(untrustedHeaders)
    foundPivot = false
    while(!foundPivot) {
        pivot := (th.Header.height + endHeight) / 2
        hp := Commit(pivot)
        if !verify(hp) { return ErrInvalidHeader(hp) }
        // try to move trusted header forward to hp
        err = CheckSupport(th,hp,trustThreshold)
        if (err != nil and err != ErrTooMuchChange) return err
        if err == nil {
            th = hp
            Store.Add(th)
            foundPivot = true
        } else {
            untrustedHeaders.add(hp)
            endHeight = pivot
        }
    }
```
or we get rid of `foundPivot`:

```
   ....
    endHeight = min(untrustedHeaders)
    while true {
        pivot := (th.Header.height + endHeight) / 2
        hp := Commit(pivot)
        if !verify(hp) { return ErrInvalidHeader(hp) }
        // try to move trusted header forward to hp
        err = CheckSupport(th,hp,trustThreshold)
        if (err != nil and err != ErrTooMuchChange) return err
        if err == nil {
            th = hp
            Store.Add(th)
            break
        } 
        untrustedHeaders.add(hp)
        endHeight = pivot
    }
```
The reason I don't like `foundPivot` is that it doesn't tell the correct story. We stay in the loop as long as we didn't find a **trusted** new higher header (pivot) and pivot is not adjacent to a trusted header. 

And along the same lines of readability:
 `if (err != nil and err != ErrTooMuchChange) return err`
is not as clear as:
 `if (err == ErrHeaderExpired or err == ErrInvalidAdjacentHeaders) return err`
or even nicer imho:
  `if fatalCheckSupportError(err) return err`